bootstrap/src/u128.ml

```diff
@@ -0,0 +1,1281 @@
+      | None -> begin
+          match ndigits with
+          | 0 -> halt "Malformed string" (* "0b" *)
+          | _ -> zero, one
+        end
+      | Some (c, i') -> begin
+          match c with
+          | '_' -> binary s i' ndigits len
+          | '0' | '1' -> begin
+              let ndigits' = Rudiments_int.succ ndigits in
+              let accum, mult = binary s i' ndigits' len in
+              let accum' = accum + (mult * (d_of_c c)) in
+              let mult' = mult * (of_usize 2) in
+              accum', mult'
+            end
+          | 'u' -> begin
+              let i'' = suffix s i' 1 len in
+              binary s i'' ndigits len
+            end
+          | _ -> halt "Malformed string"
+        end
+    end in
+    let prefix1 s i len = begin
+      match getc_opt s i len with
+      | None -> zero
+      | Some (c, i') -> begin
+          match c with
+          | 'b' -> begin
+              let accum, _ = binary s i' 0 len in
+              accum
+            end
+          | 'x' -> begin
+              let accum, _ = hexadecimal s i' 0 len in
+              accum
+            end
+          | '_' -> begin
+              let accum, _ = decimal s i' len in
+              accum
+            end
+          | '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' ->
+            let accum, mult = decimal s i' len in
+            accum + (mult * (d_of_c c))
+          | 'u' -> begin
+              let _ = suffix s i' 1 len in
+              zero
+            end
+          | _ -> halt "Malformed string"
+        end
+    end in
+    let prefix0 s i len = begin
+      let c, i' = getc s i len in
+      match c with
+      | '0' -> prefix1 s i' len
+      | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' -> begin
+          let accum, mult = decimal s i' len in
+          accum + (mult * (d_of_c c))
+        end
+      | _ -> halt "Malformed string"
+    end in
+    prefix0 s 0 (Caml.String.length s)
+
+  let c5s = of_string "0x5555_5555_5555_5555_5555_5555_5555_5555"
+  let c3s = of_string "0x3333_3333_3333_3333_3333_3333_3333_3333"
+  let c0fs = of_string "0x0f0f_0f0f_0f0f_0f0f_0f0f_0f0f_0f0f_0f0f"
+  let cff = of_string "0xff"
+
+  let bit_pop x =
+    let x =
+      x - (bit_and (bit_usr x 1) c5s) in
+    let x = (bit_and x c3s) + (bit_and (bit_usr x 2) c3s) in
+    let x = bit_and (x + (bit_usr x 4)) c0fs in
+    let x = x + (bit_usr x 8) in
+    let x = x + (bit_usr x 16) in
+    let x = x + (bit_usr x 32) in
+    let x = x + (bit_usr x 64) in
+    to_usize (bit_and x cff)
+
+  let bit_clz x =
+    let x = bit_or x (bit_usr x 1) in
+    let x = bit_or x (bit_usr x 2) in
+    let x = bit_or x (bit_usr x 4) in
+    let x = bit_or x (bit_usr x 8) in
+    let x = bit_or x (bit_usr x 16) in
+    let x = bit_or x (bit_usr x 32) in
+    let x = bit_or x (bit_usr x 64) in
+    bit_pop (bit_not x)
+
+  let bit_ctz t =
+    bit_pop (bit_and (bit_not t) (t - one))
+
+  let div_mod t0 t1 =
```

Comment:
This is the function that motivated adding `i64` and `u32`.  I initially got it working without these types, but only after much gnashing of teeth due to subtleties regarding truncation and sign extension.